src/index.js

```diff
@@ -1,174 +1,231 @@
+          socketRooms.delete(room);
+          if (socketRooms.size === 0) {
+            sids.delete(id);
+          }
+          rooms.delete(room);
+          // TODO update wild 
+          // wild.remove(room)
+        }
+      }
+      return Promise.resolve({success: true});
+    } catch (e) {
+      return Promise.reject(e);
     }
-    delete rooms[room]
-    wild.remove(room)
-    setImmediate(fn, null, null)
   }
 
   /**
    * Broadcast a packet.
-   *
-   * Options:
-   *  - `except` {Array} sids that should be excluded
-   *  - `rooms` {Array} list of rooms to broadcast to
-   *  - `method` {String} 'write' or 'send' if primus-emitter is present
-   *
-   * @param {Object} data
-   * @param {Object} opts
-   * @param {Object} clients Connected clients
+   * Used for broadcast
+   * @param {*} data - the data to transmit
+   * @param {Object} options - options to use
+   * @param {String[]} options.except - spark IDs to exclude
+   * @param {String[]} options.rooms - list of the rooms to broadcast to
+   * @param {String} options.method - 'write' or 'send' if primus-emitter is present
+   * @param {Function} options.transformer - the transformer function to call on the data
+   * @param {Object} clients -  Connected clients
    * @api public
    */
 
   function broadcast(data, options = {}, clients) {
     let rms = options.rooms || []
     if (!rms.length) {
-      return send(sids, clients, data, options)
-    }
-    for (const room of rms) {
-      let ids = rooms[room]
-      if (ids) send(ids, clients, data, options)
-      wild.match(room, key =>
-        send(rooms[key], clients, data, options)
+      let ids = new Set();
+      sids.forEach((value, key) => {
+        ids.add(key);
+      });
+      return send(ids, clients, data, options)
     }
+    let promises = [];
+    rooms.forEach((value, room) => {
+      let ids = rooms.get(room);
+      if (ids instanceof Set) {
+        promises.push(send(ids, clients, data, options));
+      }
+      // TODO wild support
+      // wild.match(room, key =>
+      //   send(rooms[key], clients, data, options)
+      // )
+    });
+    return Promise.all(promises);
   }
 
   /**
    * Create sender.
-   *
-   * @param {Array} clients
-   * @param {Mixed} data
-   * @param {Object} options
+   * Helper for broadcast
+   * @param {Set} ids - spark IDs to send to
+   * @param {Object} clients - connected clients
+   * @param {*} data - the data to send
+   * @param {Object} options - same options as broadcast
    * @api private
    */
 
   function send(ids, clients, data, options) {
-    const { except = [], method = 'writer', transformer } = options
-    let sent = {}
-    for (let id in ids) {
-      const socket = clients[id]
-      if (sent[id] || ~except.indexOf(id) || !socket) continue
-      transform(socket, data, method, transformer)
-      sent[id] = true
+    const { except = [], method = 'write', transformer } = options
+    let promises = [];
+    try {
```

Comment:
Ok for this one i remember now why i have the `try...catch` it is there in case the transformer function throws due to custom logic passed in